rivertile/main.zig

```diff
@@ -14,129 +14,322 @@
+                            width,
+                            height,
+                        ),
+                        .top => layout.pushViewDimensions(
+                            data.serial,
+                            y,
+                            x,
+                            height,
+                            width,
+                        ),
+                        .bottom => layout.pushViewDimensions(
+                            data.serial,
+                            y,
+                            @intCast(i32, usable_width - width) - x,
+                            height,
+                            width,
+                        ),
+                    }
+                }
+
+                // Layout pseudo-symbols are taken from dwm
+                layout.commit(data.serial, switch (self.orientation) {
+                    .left => "[]=",
+                    .right => "=[]",
+                    .top => "TTT",
+                    .bottom => "LLL",
+                });
+            },
+            .advertise_view => {},
+            .advertise_done => {},
         }
     }
+};
 
-    try stdout_buf.flush();
-}
+pub fn main() !void {
+    const display = wl.Display.connect(null) catch {
+        std.debug.warn("Unable to connect to Wayland server.\n", .{});
+        std.os.exit(1);
+    };
+    defer display.disconnect();
+
+    var context: Context = .{};
+
+    const args = std.os.argv;
+    if (args.len == 2) {
+        context.view_padding = try std.fmt.parseInt(u32, std.mem.spanZ(args[1]), 10);
+    }
 
-fn printUsageAndExit() noreturn {
-    const usage: []const u8 =
-        \\Usage: rivertile left|right|top|bottom [args passed by river]
-        \\
-    ;
+    const registry = try display.getRegistry();
+    try registry.setListener(*Context, registryListener, &context);
+    _ = try display.roundtrip();
 
-    std.debug.warn(usage, .{});
-    std.os.exit(1);
+    if (context.layout_manager == null) {
+        std.debug.warn("Wayland server does not support zriver_layout_unstable_v1.\n", .{});
+        std.os.exit(1);
+    }
+
+    context.configureAllOutputs();
+    defer context.destroyAllOutputs();
+
+    while (context.running) {
+        _ = display.dispatch() catch {
+            context.running = false;
+        };
+    }
+}
+
+fn registryListener(registry: *wl.Registry, event: wl.Registry.Event, context: *Context) void {
+    switch (event) {
+        .global => |global| {
+            if (std.cstr.cmp(global.interface, zriver.LayoutManagerV1.getInterface().name) == 0) {
+                context.layout_manager = registry.bind(global.name, zriver.LayoutManagerV1, 1) catch return;
+            } else if (std.cstr.cmp(global.interface, wl.Output.getInterface().name) == 0) {
+                context.addOutput(registry, global.name) catch {
+                    std.debug.warn("Failed to bind output.\n", .{});
+                    context.running = false;
+                };
```

Comment:
This doesn't handle the case where an output is advertised before the river layout manager properly. Currently `addOutput()` will return an error and we will exit.

Reply:
I am pretty sure it does. If the configuration fails due to the river layout manager not being available, it will not error out. I have edited the code to make this more clear.